src/tests/end2end/DepthStencilStateTests.cpp

```diff
@@ -0,0 +1,512 @@
+        // (the base triangle is always drawn first)
+        // Apply the given stencil comparison function to the triangle
+        // The less, equal, greater booleans denote wether the respective triangle should be visible based on the comparison function
+        void CheckStencilCompareFunction(nxt::CompareFunction compareFunction, bool less, bool equal, bool greater) {
+            nxt::DepthStencilState baseState = device.CreateDepthStencilStateBuilder()
+                .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Always, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Replace)
+                .GetResult();
+
+            nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+                .SetStencilFunction(nxt::Face::Both, compareFunction, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep)
+                .GetResult();
+
+            RGBA8 baseColor = RGBA8(255, 255, 255, 255);
+            RGBA8 lessColor = RGBA8(255, 0, 0, 255);
+            RGBA8 equalColor = RGBA8(0, 255, 0, 255);
+            RGBA8 greaterColor = RGBA8(0, 0, 255, 255);
+
+            TestSpec base = { baseState, baseColor, 0.0f, 1u };
+            DoTest({ base, { state, lessColor, 0.f, 0u } }, less ? lessColor : baseColor);
+            DoTest({ base, { state, equalColor, 0.f, 1u } }, equal ? equalColor : baseColor);
+            DoTest({ base, { state, greaterColor, 0.f, 2u } }, greater ? greaterColor : baseColor);
+        }
+
+        // Wipe the stencil value to `initialStencil` then draw a triangle such that is always passes the depth and stencil tests with the provided reference
+        // Check that the stencil value is as expected
+        void CheckStencilOperation(nxt::StencilOperation stencilOperation, uint32_t initialStencil, uint32_t reference, uint32_t expectedStencil) {
+            nxt::DepthStencilState baseState = device.CreateDepthStencilStateBuilder()
+                .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Always, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Replace)
+                .GetResult();
+
+            nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+                .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Always, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, stencilOperation)
+                .GetResult();
+
+            CheckStencil({
+                { baseState, RGBA8(255, 255, 255, 255), 0.f, initialStencil },
+                { state, RGBA8(255, 0, 0, 255), 0.f,  reference },
+            }, expectedStencil);
+        }
+
+        // Given a list of test specs, add an additional test that always draws a green triangle if the stencil value is equal to the expected value
+        void CheckStencil(const std::vector<TestSpec> &testParams, uint32_t expectedStencil) {
```

Comment:
nit: You could simplify this function a little bit by having it take the vector by value.